src/computability/regex.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Fox Thomson.
+-/
+import data.fintype.basic
+import data.finset.basic
+import tactic.rcases
+import tactic.omega
+
+/-!
+# Regular Expression
+
+This file contains the formal definition for regular expressions on basic lemmas.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- This is the definition of regular expressions
+  `RZero` matches nothing
+  `RNull` matches only the empty string
+  `RChar a` matches only the string 'a'
+  `RStar M` matches any finite concatenation of strings which match `M`
+  `RPlus M N` matches anything which match `M` or `N`
+  `RComp M N` matches `x ++ y` if `x` matches `M` and `y` matches `N` -/
+inductive regex (α : Type u) : Type (u+1)
+| RZero : regex
+| RNull : regex
+| RChar : α → regex
+| RStar : regex → regex
+| RPlus : regex → regex → regex
+| RComp : regex → regex → regex
+
+namespace regex
+
+instance regex_inhabited : inhabited (regex α) := ⟨ RZero ⟩
+
+/-- `match_null M` is true if and only if `M` matches the empty string -/
+def match_null : regex α → bool
+| RZero := ff
+| RNull := tt
+| (RChar _) := ff
+| (RStar M) := tt
+| (RPlus M N) := M.match_null || N.match_null
+| (RComp M N) := M.match_null && N.match_null
+
+include dec
+
+/-- `M.feed a` matches `x` if `M` matches `"a" ++ x` -/
+def feed : regex α → α → regex α
+| RZero _ := RZero
+| RNull _ := RZero
+| (RChar a₁) a₂ := ite (a₁ = a₂) RNull RZero
```

Comment:
Note this can also be written with `if ... then ... else`, but up to you if you prefer that:
```suggestion
| (RChar a₁) a₂ := if a₁ = a₂ then RNull else RZero
```